File: mcp/query_tenant/src/client.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { AuthenticatedClient } from './client.js';
import type { SessionCredentials, AkskCredentials } from './auth.js';

const sessionCreds: SessionCredentials = {
  mode: 'session',
  token: 'Bearer test-session-token',
  headerName: 'Authorization',
};

const akskCreds: AkskCredentials = {
  mode: 'aksk',
  accessKey: 'AK_TEST',
  secretKey: 'SK_TEST',
};

describe('AuthenticatedClient', () => {
  describe('constructor', () => {
    it('rejects non-HTTPS base URL by default', () => {
      expect(() => new AuthenticatedClient('http://insecure.com', sessionCreds))
        .toThrow('non-HTTPS');
    });

    it('allows HTTP when allowHttp=true', () => {
      expect(() => new AuthenticatedClient('http://internal.local', sessionCreds, true))
        .not.toThrow();
    });

    it('accepts HTTPS base URL', () => {
      expect(() => new AuthenticatedClient('https://secure.com', sessionCreds))
        .not.toThrow();
    });
  });

  describe('request with session auth', () => {
    let fetchSpy: ReturnType<typeof vi.fn>;

    beforeEach(() => {
      fetchSpy = vi.fn();
      vi.stubGlobal('fetch', fetchSpy);
    });

    afterEach(() => {
      vi.restoreAllMocks();
    });

    it('injects session token header on GET', async () => {
      fetchSpy.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: () => Promise.resolve({ result: 'ok' }),
      });

      const client = new AuthenticatedClient('https://api.example.com', sessionCreds);
      const res = await client.request({ path: '/data', method: 'GET' });

      expect(res.status).toBe(200);
      expect(res.data).toEqual({ result: 'ok' });

      const [url, opts] = fetchSpy.mock.calls[0];
      expect(url).toContain('https://api.example.com/data');
      expect(opts.headers['Authorization']).toBe('Bearer test-session-token');
    });

    it('does NOT expose token in request body', async () => {
      fetchSpy.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: () => Promise.resolve({}),
      });

      const client = new AuthenticatedClient('https://api.example.com', sessionCreds);
      await client.request({ path: '/data', method: 'POST', body: { query: 'test' } });

      const [, opts] = fetchSpy.mock.calls[0];
      expect(opts.body).not.toContain('test-session-token');
    });
  });

  describe('request with aksk auth', () => {
    let fetchSpy: ReturnType<typeof vi.fn>;

    beforeEach(() => {
      fetchSpy = vi.fn();
      vi.stubGlobal('fetch', fetchSpy);
    });

    afterEach(() => {
      vi.restoreAllMocks();
    });

    it('injects HMAC signature headers', async () => {
      fetchSpy.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: () => Promise.resolve({ data: [1, 2, 3] }),
      });

      const client = new AuthenticatedClient('https://api.example.com', akskCreds);
      await client.request({ path: '/data' });

      const [, opts] = fetchSpy.mock.calls[0];
      expect(opts.headers['X-Access-Key']).toBe('AK_TEST');
      expect(opts.headers['X-Timestamp']).toBeDefined();
      expect(opts.headers['X-Signature']).toMatch(/^[0-9a-f]{64}$/);
    });

    it('does NOT include secret key in any header', async () => {
      fetchSpy.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: () => Promise.resolve({}),
      });

      const client = new AuthenticatedClient('https://api.example.com', akskCreds);
      await client.request({ path: '/test' });

      const [, opts] = fetchSpy.mock.calls[0];
      const allHeaderValues = Object.values(opts.headers as Record<string, string>).join(' ');
      expect(allHeaderValues).not.toContain('SK_TEST');
    });
  });

  describe('error handling', () => {
    let fetchSpy: ReturnType<typeof vi.fn>;

    beforeEach(() => {
      fetchSpy = vi.fn();
      vi.stubGlobal('fetch', fetchSpy);
    });

    afterEach(() => {
      vi.restoreAllMocks();
    });

    it('returns structured error on HTTP 4xx/5xx', async () => {
      fetchSpy.mockResolvedValue({
        ok: false,
        status: 403,
        json: () => Promise.resolve({ reason: 'forbidden' }),
      });

      const client = new AuthenticatedClient('https://api.example.com', sessionCreds);
      const res = await client.request({ path: '/secret' });

      expect(res.status).toBe(403);
      expect((res.data as Record<string, unknown>).error).toBe(true);
    });

    it('retries on network failure and returns 502 after exhaustion', async () => {
      fetchSpy.mockRejectedValue(new Error('network error'));

      const client = new AuthenticatedClient('https://api.example.com', sessionCreds);
      const res = await client.request({ path: '/fail', timeoutMs: 100 });

      expect(res.status).toBe(502);
      expect((res.data as Record<string, unknown>).message).toContain('All 3 attempts failed');
      expect(fetchSpy).toHaveBeenCalledTimes(3);
    }, 30000);

    it('trips circuit breaker after repeated failures', async () => {
      fetchSpy.mockRejectedValue(new Error('down'));

      const client = new AuthenticatedClient('https://api.example.com', sessionCreds);

      for (let i = 0; i < 2; i++) {
        await client.request({ path: '/fail', timeoutMs: 100 });
      }

      await expect(client.request({ path: '/fail', timeoutMs: 100 }))
        .rejects.toThrow('Circuit breaker is OPEN');
    }, 60000);
  });

  describe('timeout', () => {
    afterEach(() => {
      vi.restoreAllMocks();
    });

    it('aborts request after timeout', async () => {
      const fetchSpy = vi.fn().mockImplementation(
        () => new Promise((_, reject) => {
          setTimeout(() => reject(new DOMException('The operation was aborted', 'AbortError')), 50);
        })
      );
      vi.stubGlobal('fetch', fetchSpy);

      const client = new AuthenticatedClient('https://api.example.com', sessionCreds);
      const res = await client.request({ path: '/slow', timeoutMs: 50 });

      expect(res.status).toBe(502);
    }, 30000);
  });
});

File: mcp/query_tenant_go/client_test.go
package main

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func newSessionCreds() *Credentials {
	return &Credentials{
		Mode:    AuthModeSession,
		Session: &SessionCredentials{Token: "Bearer test-session-token", HeaderName: "Authorization"},
	}
}

func newAKSKCreds() *Credentials {
	return &Credentials{
		Mode: AuthModeAKSK,
		AKSK: &AKSKCredentials{AccessKey: "AK_TEST", SecretKey: "SK_TEST"},
	}
}

func TestClient_RejectsHTTP(t *testing.T) {
	_, err := NewAuthenticatedClient("http://insecure.com", newSessionCreds(), false)
	if err == nil {
		t.Error("expected error for non-HTTPS URL")
	}
}

func TestClient_AllowsHTTP(t *testing.T) {
	_, err := NewAuthenticatedClient("http://internal.local", newSessionCreds(), true)
	if err != nil {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestClient_AcceptsHTTPS(t *testing.T) {
	_, err := NewAuthenticatedClient("https://secure.com", newSessionCreds(), false)
	if err != nil {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestClient_SessionAuth(t *testing.T) {
	var gotAuth string
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("Authorization")
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{"result": "ok"})
	}))
	defer ts.Close()

	client, err := NewAuthenticatedClient(ts.URL, newSessionCreds(), true)
	if err != nil {
		t.Fatal(err)
	}

	resp, err := client.Request(QueryOptions{Path: "/data", Method: "GET"})
	if err != nil {
		t.Fatal(err)
	}

	if resp.Status != 200 {
		t.Errorf("status = %d, want 200", resp.Status)
	}
	if gotAuth != "Bearer test-session-token" {
		t.Errorf("auth header = %q", gotAuth)
	}
}

func TestClient_AKSKAuth(t *testing.T) {
	var gotAK, gotSig, gotTS string
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAK = r.Header.Get("X-Access-Key")
		gotSig = r.Header.Get("X-Signature")
		gotTS = r.Header.Get("X-Timestamp")
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{"data": []int{1, 2, 3}})
	}))
	defer ts.Close()

	client, err := NewAuthenticatedClient(ts.URL, newAKSKCreds(), true)
	if err != nil {
		t.Fatal(err)
	}

	_, err = client.Request(QueryOptions{Path: "/data"})
	if err != nil {
		t.Fatal(err)
	}

	if gotAK != "AK_TEST" {
		t.Errorf("X-Access-Key = %q", gotAK)
	}
	if gotTS == "" {
		t.Error("X-Timestamp is empty")
	}
	if len(gotSig) != 64 {
		t.Errorf("X-Signature length = %d, want 64", len(gotSig))
	}
}

func TestClient_AKSKNoSecretInHeaders(t *testing.T) {
	var allHeaders http.Header
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		allHeaders = r.Header
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{})
	}))
	defer ts.Close()

	client, err := NewAuthenticatedClient(ts.URL, newAKSKCreds(), true)
	if err != nil {
		t.Fatal(err)
	}

	_, err = client.Request(QueryOptions{Path: "/test"})
	if err != nil {
		t.Fatal(err)
	}

	for k, vals := range allHeaders {
		for _, v := range vals {
			if v == "SK_TEST" {
				t.Errorf("secret key exposed in header %q", k)
			}
		}
	}
}

func TestClient_HTTP4xx(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(403)
		json.NewEncoder(w).Encode(map[string]any{"reason": "forbidden"})
	}))
	defer ts.Close()

	client, err := NewAuthenticatedClient(ts.URL, newSessionCreds(), true)
	if err != nil {
		t.Fatal(err)
	}

	resp, err := client.Request(QueryOptions{Path: "/secret"})
	if err != nil {
		t.Fatal(err)
	}

	if resp.Status != 403 {
		t.Errorf("status = %d, want 403", resp.Status)
	}
	data := resp.Data.(map[string]any)
	if data["error"] != true {
		t.Error("expected error=true")
	}
}

func TestClient_RetryExhaustion(t *testing.T) {
	callCount := 0
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		callCount++
		hj, ok := w.(http.Hijacker)
		if ok {
			conn, _, _ := hj.Hijack()
			conn.Close()
			return
		}
		w.WriteHeader(500)
	}))
	defer ts.Close()

	client, err := NewAuthenticatedClient(ts.URL, newSessionCreds(), true)
	if err != nil {
		t.Fatal(err)
	}

	resp, err := client.Request(QueryOptions{Path: "/fail", TimeoutMs: 1000})
	if err != nil {
		if resp == nil {
			t.Fatalf("expected 502 response, got error: %v", err)
		}
	}

	if resp != nil && resp.Status != 502 {
		t.Errorf("status = %d, want 502", resp.Status)
	}
	if callCount < 2 {
		t.Errorf("expected at least 2 calls, got %d", callCount)
	}
}
